internal/jobcontainers/job.go

```diff
@@ -0,0 +1,308 @@
+	if err != nil {
+		return err
+	}
+	defer windows.Close(hProc)
+	return windows.AssignProcessToJobObject(job.jobHandle, hProc)
+}
+
+// Terminates the job, essentially calls TerminateProcess on every process in the
+// job.
+func (job *jobObject) terminate() error {
+	if job.jobHandle != 0 {
+		return windows.TerminateJobObject(job.jobHandle, 1)
+	}
+	return nil
+}
+
+func (job *jobObject) shutdown(ctx context.Context) error {
+	pids, err := job.pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	var (
+		terminate bool
+		signalErr bool
+	)
+	for _, pid := range pids {
+		if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, pid); err != nil {
+			log.G(ctx).WithField("pid", pid).Error("failed to send ctrl-break to process in job")
+			signalErr = true
+		}
+	}
+
+	// Get pids in job again. If there is any left then terminate the job.
+	newPids, err := job.pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	terminate = len(newPids) != 0
+	// If any of the processes couldnt be killed gracefully just terminate the job.
+	// Equivalent to calling TerminateProcess on every proc in the job.
+	if terminate || signalErr {
+		return job.terminate()
+	}
+	return nil
+}
+
+// Returns all of the process IDs in the job object.
+func (job *jobObject) pids() ([]uint32, error) {
+	info := winapi.JOBOBJECT_BASIC_PROCESS_ID_LIST{}
+	var retLen uint32
+	err := winapi.QueryInformationJobObject(
+		job.jobHandle,
+		winapi.JobObjectBasicProcessIdList,
+		uintptr(unsafe.Pointer(&info)),
+		uint32(unsafe.Sizeof(info)),
+		&retLen,
+	)
+
+	// If there are no processes in the job the next call to QueryInformation will just
+	// hang until a memory alloc error. Return here.
+	if err == nil {
+		// Return empty slice instead of nil to play well with the caller of this.
+		// I don't think we should return an error on there being no PIDs.
+		var pids []uint32
+		return pids, nil
+	}
+
+	if err != nil && err != winapi.ERROR_MORE_DATA {
+		return nil, fmt.Errorf("failed initial query for PIDs in job object: %s", err)
+	}
+
+	buf := make([]uintptr, unsafe.Sizeof(info)+unsafe.Sizeof(info.ProcessIdList[0])*uintptr(info.NumberOfAssignedProcesses-1))
+	err = winapi.QueryInformationJobObject(
+		job.jobHandle,
+		winapi.JobObjectBasicProcessIdList,
+		uintptr(unsafe.Pointer(&buf[0])),
+		uint32(len(buf)),
+		&retLen)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to query for PIDs in job object: %s", err)
+	}
+
+	bufInfo := (*winapi.JOBOBJECT_BASIC_PROCESS_ID_LIST)(unsafe.Pointer(&buf[0]))
+	rawPids := make([]uintptr, bufInfo.NumberOfProcessIdsInList)
+
+	err = winapi.RtlMoveMemory(
+		(*byte)(unsafe.Pointer(&rawPids[0])),
+		(*byte)(unsafe.Pointer(&bufInfo.ProcessIdList[0])),
+		uintptr(bufInfo.NumberOfProcessIdsInList)*unsafe.Sizeof(rawPids[0]))
+	if err != nil {
+		return nil, fmt.Errorf("failed to move PID info to new buffer: %s", err)
+	}
```

Comment:
why do we need to move to a new buffer?

Reply:
This function is a mess. I'll refactor tonight, I did this a month or so ago and the knowledge has left me on why most of this was needed to work

Reply:
We move to a new buffer with a length of the number of processes in the job so we have access to the other pids (if there's more than one) as we have `ProcessIdList` defined as `[1]uintptr` to match the definition https://docs.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-jobobject_basic_process_id_list. We could either change `[1]uintptr` to something arbitrarily large that we imagine we'd never hit (100, 1000 etc) or do some ugly go cast that will basically do what this call is doing anyways.